modules/core/src/main/java/org/apache/ignite/spi/encryption/keystore/KeystoreEncryptionSpi.java

```diff
@@ -0,0 +1,502 @@
+        }
+    });
+
+    /** */
+    private ThreadLocal<Cipher> aesWithoutPadding = ThreadLocal.withInitial(() -> {
+        try {
+            return Cipher.getInstance(AES_WITHOUT_PADDING);
+        }
+        catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+            throw new IgniteException(e);
+        }
+    });
+
+    /** {@inheritDoc} */
+    @Override public void spiStart(@Nullable String igniteInstanceName) throws IgniteSpiException {
+        assertParameter(!F.isEmpty(keyStorePath), "KeyStorePath shouldn't be empty");
+        assertParameter(keyStorePwd != null && keyStorePwd.length > 0,
+            "KeyStorePassword shouldn't be empty");
+
+        try (InputStream keyStoreFile = keyStoreFile()) {
+            assertParameter(keyStoreFile != null, keyStorePath + " doesn't exists!");
+
+            KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
+
+            ks.load(keyStoreFile, keyStorePwd);
+
+            if (log != null)
+                log.info("Successfully load keyStore [path=" + keyStorePath + "]");
+
+            masterKey = new KeystoreEncryptionKey(ks.getKey(masterKeyName, keyStorePwd), null);
+        }
+        catch (GeneralSecurityException | IOException e) {
+            throw new IgniteSpiException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void spiStop() throws IgniteSpiException {
+        ensureStarted();
+
+        //empty.
+    }
+
+    /** {@inheritDoc} */
+    @Override public byte[] masterKeyDigest() {
+        ensureStarted();
+
+        return makeDigest(masterKey.key().getEncoded());
+    }
+
+    /** {@inheritDoc} */
+    @Override public KeystoreEncryptionKey create() throws IgniteException {
+        ensureStarted();
+
+        try {
+            KeyGenerator gen = KeyGenerator.getInstance(CIPHER_ALGO);
+
+            gen.init(keySize);
+
+            SecretKey key = gen.generateKey();
+
+            return new KeystoreEncryptionKey(key, makeDigest(key.getEncoded()));
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new IgniteException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void encrypt(ByteBuffer data, Serializable key, ByteBuffer res) {
+        doEncryption(data, aesWithPadding.get(), key, res);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void encryptNoPadding(ByteBuffer data, Serializable key, ByteBuffer res) {
+        doEncryption(data, aesWithoutPadding.get(), key, res);
+    }
+
+    /** {@inheritDoc} */
+    @Override public byte[] decrypt(byte[] data, Serializable key) {
+        assert key instanceof KeystoreEncryptionKey;
+
+        ensureStarted();
+
+        try {
+            SecretKeySpec keySpec = new SecretKeySpec(((KeystoreEncryptionKey)key).key().getEncoded(), CIPHER_ALGO);
+
+            Cipher cipher = aesWithPadding.get();
+
+            cipher.init(DECRYPT_MODE, keySpec, new IvParameterSpec(data, 0, cipher.getBlockSize()));
```

Comment:
I think it can be not secure to use data as IV and as data at the same time because XOR can give us plaintext data

Reply:
We doesn't use data as IV.
IV just a first 16 bytes in page.
It's a common practice to do so.

Please, take a look at `encrypt` method.

Reply:
yes, my fault, I've just missed that data is used with different offsets.